rust/datafusion/src/logical_plan/plan.rs

```diff
@@ -94,6 +101,21 @@ pub enum LogicalPlan {
         /// The incoming logical plan
         input: Arc<LogicalPlan>,
     },
+    /// Join two logical plans on one or more join columns
+    Join {
+        /// Left input
+        left: Arc<LogicalPlan>,
+        /// Right input
+        right: Arc<LogicalPlan>,
+        /// Columns in left input to use for join keys
+        left_keys: Vec<String>,
+        /// Columns in right input to use for join keys
+        right_keys: Vec<String>,
```

Comment:
I would use `Vec<(left_i, right_i)>` because it automatically enforces the invariant that `left_keys.len() == right_keys.len()`. We can still keep the public interface `left_keys, right_keys` and perform the check before passing them to the builder.

Atm, when we use `left.zip(right)`, we take the shortest vector, which may hide a bug in the code.

Reply:
Thanks. I agree and I have made this change. The user-facing DataFrame method is now the only plan that accepts the two separate lists of column names and we verify they are the same length when creating the logical plan.